mlir-tensorrt/compiler/include/mlir-tensorrt/Compiler/OptionsRegistry.h

```diff
@@ -0,0 +1,85 @@
+//-*-===//
+//
+// SPDX-FileCopyrightText: Copyright 2024 NVIDIA CORPORATION & AFFILIATES.
+// All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+///
+/// This is the top-level C++ interface for clients that wish to compile
+/// MLIR programs from a supported input (e.g. StableHLO) into a supported
+/// target (e.g. TensorRT engine). This file contains just the declarations for
+/// the CompilerClient object, see below for details.
+///
+//===----------------------------------------------------------------------===//
+#ifndef MLIR_TENSORRT_COMPILER_OPTIONS_REGISTRY
+#define MLIR_TENSORRT_COMPILER_OPTIONS_REGISTRY
+
+#include "Client.h"
+#include "mlir-tensorrt-dialect/Utils/Options.h"
+#include "mlir-tensorrt/Dialect/Plan/IR/Plan.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/StringRef.h"
+#include <functional>
+
+namespace mlirtrt::compiler {
+
+using OptionsConstructorFuncT =
+    std::function<StatusOr<std::unique_ptr<mlir::OptionsContext>>(
+        const CompilerClient &client, const llvm::ArrayRef<llvm::StringRef>)>;
+
+void registerOption(const llvm::StringRef optionsType,
+                    OptionsConstructorFuncT func);
+
+StatusOr<std::unique_ptr<mlir::OptionsContext>>
+createOptions(const CompilerClient &client, const llvm::StringRef optionsType,
+              const llvm::ArrayRef<llvm::StringRef> args);
+
+/// Helper to build callbacks that can create options.
+template <typename OptionsT, typename TaskT>
+StatusOr<std::unique_ptr<mlir::OptionsContext>>
+optionsCreateFromArgs(const CompilerClient &client,
+                      const llvm::ArrayRef<llvm::StringRef> args) {
+  // Load available extensions.
+  mlir::MLIRContext *context = client.getContext();
+  mlir::plan::PlanDialect *planDialect =
+      context->getLoadedDialect<mlir::plan::PlanDialect>();
+  compiler::TaskExtensionRegistry extensions =
+      planDialect->extensionConstructors.getExtensionRegistryForTask<TaskT>();
+
+  auto result = std::make_unique<OptionsT>(std::move(extensions));
+
+  std::string err;
+  if (failed(result->parse(args, err))) {
+    return getInternalErrorStatus(
+        "failed to parse options string \"{0:$[ ]}\" due to error {1}",
+        llvm::iterator_range(args), err);
+  }
+
+  // TODO: Figure out whether to add a method in the base class like
```

Comment:
OK I see what you are saying. The intended behavior is that in the constructor of the options object, it should check whether `shouldInferDeviceOptionsFromHost` is set. If so, then invoke this method. Otherwise, don't invoke this method.

Reply:
That makes sense. As we discussed in the meeting I can make this change in the next PR where I will implement the TRT options/task.